### PR TITLE
deprecated erlang:get_stacktrace/0 function

### PR DESCRIPTION
**Pull request by getong**

### CODE REVIEW

**Comment by michalwski** [comment]
I'd really love to see a short description to the PR so that all the changes are clearly summarised.

**Comment by getong** [comment]
@michalwski New msg updated.

**Comment by getong** [comment]
@michalwski 
As erlang/otp 21 is released, a few functions are deprecated and need to write some code.
I notice the MongooseIM use the [worker_pool](https://github.com/inaka/worker_pool), I also made a pr to it. But the worker_pool only supports erlang/otp 21, you can see the inaka/worker_pool#138 , that is made by me.
If you want to support erlang version from 17 to 21, you can fork the worker_pool repo, and I can make some code to it.

**Comment by michalwski** [comment]
There is no need to fork worker_pool. We use `red-3.1` from the repo which compile and runs on all Erlang/OTP versions we support. Currently these are 19, 20 and 21.

**Comment by getong** [comment]
@michalslaski Don' merge it, something needs to fix.

File: include/mongoose.hrl [hunk 1]
```diff
@@ -35,4 +35,10 @@
 %% Logging mechanism
 -include("mongoose_logger.hrl").
 
-
+-ifdef(OTP_RELEASE). %% this implies 21 or higher
```
Comment (fenek) [inline include/mongoose.hrl]: Where is it defined? Besides, it's not a good name, as it doesn't imply OTP 21 or higher.

File: src/auth/ejabberd_auth_odbc.erl [hunk 2]
```diff
@@ -128,11 +128,10 @@ check_password(LUser, LServer, Password, Digest, DigestGen) ->
                        "reason=~p user=~ts", [Error, LUser]),
             false %% Typical error is that table doesn't exist
     catch
-        Class:Reason ->
-            Stacktrace = erlang:get_stacktrace(),
+        ?EXCEPTION(Class, Reason, CURRENT_STACKTRACE) ->
```
Comment (fenek) [inline src/auth/ejabberd_auth_odbc.erl]: Can we make it just `Stacktrace`? After preprocessing it's an ordinary variable after all.

**Comment by getong** [comment]
@fenek 
http://www.erlang.org/news/123
```
A new predifined macro `OTP_RELEASE`
```

**Comment by getong** [comment]
@fenek  @michalwski 
I rewrite the code, feedback is needed.

File: include/mongoose.hrl (hunk 1 above)
Comment (erszcz) [reply include/mongoose.hrl]: From [_Erlang OTP 21.0 is released_](http://www.erlang.org/news/123) (original spelling):

> A new predifined macro `OTP_RELEASE` and preprocessor directives `-if` and `-elif`

If it was added in OTP 21 we can be pretty sure it implies OTP 21 or higher.

Comment (kzemek) [reply include/mongoose.hrl]: It also [has a value](http://erlang.org/doc/reference_manual/macros.html) so we could make it more explicit

**Comment by fenek** [comment]
Hi @getong 

Together with the team we've decided that a lack of warning is not a sufficient benefit at a cost of worse looking source code. If these deprecation warnings cause you any problems, maybe using `{nowarn_deprecated_function, {erlang, get_stacktrace, 0}}` compiler option in `mongoose.hrl` is a less invasive choice for now? The old method to get stack trace works in OTP 21 anyway.

**Comment by getong** [comment]
@fenek 
I don't think so. 
[EEP 47: Add syntax in try/catch to retrieve the stacktrace directly](https://www.erlang.org/erlang-enhancement-proposals/eep-0047.html)
```
The main motivation for this feature is to be able to deprecate (and ultimately remove) erlang:get_stacktrace/0.

The problem with erlang:get_stacktrace/0 is that it forces the stacktrace from the latest exception in a process to be retained until another exception occurs or the process terminates. 
The stacktrace often includes the arguments for the last function call, BIF call, or (in OTP 21) operator that failed. 
The arguments can be of any size.
```
And the macro is copied from [meck](https://github.com/eproxus/meck)
You can see it at [the code](https://github.com/eproxus/meck/blob/master/src/meck_code_gen.erl#L186-L192)

**Comment by getong** [comment]
The build failed reason is the spec wrong, I fix it and made a new pr, see #2010

**Comment by fenek** [comment]
@getong 

Thank you, I wasn't aware of this difference. Then it is indeed an improvement.

**Comment by fenek** [comment]
Thank you for your effort but we've decided to do a proper migration when MongooseIM stops supporting OTP versions older than 21. The optimisation described in EEP won't be possible until `get_stacktrace` is removed completely.

**Pull request closed without merging**